feat(logs): Rate limiter

benjackwhite opened the pull request:
## Problem

We want to rate limit incoming data requests. In order to do this we will set a global per-team limit for processing.

## Changes

- [x] Adds a rate limiter lua v2 that returns the info needed to know how many tokens should have been allowed
- [x] Implements the limiter call in the processor
- [x] Implement team check so we only do it for one team (so we can rate limit ourselves first and check it works)
- [x] Add prometheus metrics for tracking number of messages dropped

## Follow up 
- [ ] Uses the timestamp of the kafka message and not the clock time to account for ingestion delays
- [ ] Add appmetrics2 tracking for rates and drop rates etc which will allow reporting in the UI

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

<!-- Docs reminder: If this change requires updated docs, please do that! Engineers are the primary people responsible for their documentation. 🙌 -->

👉 _Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review._

## Changelog: (features only) Is this feature complete?

<!-- Yes if this is okay to go in the changelog. No if it's still hidden behind a feature flag, or part of a feature that's not complete yet, etc.  -->
<!-- Removing this section does not mean the changelog bot won't pick it up, because *some people* like to not use the template, so we can't rely on it existing. -->

benjackwhite committed:
Move redis to generic place

benjackwhite committed:
fix

benjackwhite committed:
Fixes

benjackwhite committed:
Metrics and filtering based on limits

benjackwhite committed:
Fixes

frankh commented on plugin-server/src/logs-ingestion/services/logs-rate-limiter.service.ts:
```diff
@@ -0,0 +1,135 @@
+                continue
+            }
+            const currentCost = teamCosts.get(message.teamId) ?? 0
+            // Cost is in KB (uncompressed bytes / 1024)
```
i wonder if we should use KB/MB instead of KiB/MiB

I feel like more and more the SI units are standard and base 2 is dying, might be more obvious for customers?

Then again, base2 is more generous and we might get complaints about throttling at 0.95MiB

benjackwhite committed:
Fix

benjackwhite committed:
fix

benjackwhite committed:
fix

frankh committed:
KB not KiB

frankh committed:
add tls config to logs redis

frankh committed:
fix tls config to be an object

frankh committed:
add default config

frankh committed:
add default value for logs ingestion redis tls

jonmcwest committed:
add LOGS_LIMITER_DISABLED_FOR_TEAMS to config and update rate limiting logic

jonmcwest committed:
add team-specific logs rate limiting configuration and update service logic

jonmcwest committed:
perf(logs-rate-limiter): parse enabled/disabled team lists at construction time

Move parsing of LOGS_LIMITER_ENABLED_TEAMS and LOGS_LIMITER_DISABLED_FOR_TEAMS
from isRateLimitingEnabledForTeam() into the constructor. Uses Set for O(1)
lookups instead of O(n) array scanning on every call.

jonmcwest commented on plugin-server/src/logs-ingestion/services/logs-rate-limiter.service.ts:
```diff
@@ -70,12 +70,30 @@ export class LogsRateLimiterService {
+                .includes(teamId)
+        ) {
+            return false
+        }
```
I've moved this into the constructor with the custom team bucket parsing so it's calculated once [08ebd2e](https://github.com/PostHog/posthog/pull/41888/commits/08ebd2ef1469db4ed872bf0a17c2825aa5920325)
